SPEC.md

```diff
@@ -0,0 +1,115 @@
+# libp2p Daemon
+
+The libp2p daemon is a standalone binary meant to make it easy to bring
+peer-to-peer networking to new languages without fully porting libp2p and all
+of its complexities.
+
+_At the moment, this is a living document. As such, it will be susceptible to
+changes until stabilization._
+
+## Structure
+
+### Overview
+
+There are two pieces to the libp2p daemon:
+
+- __Daemon__: A golang daemon that manages libp2p hosts and proxies streams to
+  the end user.
+- __Client__: A library written in any language that controls the daemon over
+  a protocol specified in this document, allowing end users to enjoy the
+  benefits of peer-to-peer networking without implementing a full libp2p stack.
+
+### Technical Details
+
+The libp2p daemon and client will communicate with each other over an HTTP API
+for the time being. In the future, this may likely evolve into a simpler,
+lighter weight protocol. Both the daemon and the client will run HTTP servers,
+facilitating bidirectional communication.
+
+In the initial implementation, communication between end users and the daemon
+will take place over unix sockets. Each unix socket will correspond to a libp2p
+stream, operating on a given protocol.
+
+Future implementations may attempt to take advantage of shared memory (shmem)
+or other IPC constructs.
+
+## HTTP API Specification
+
+### Data Types
+
+_Keys surrounded in <> brackets are optional. Values surrounded in [] are
+arrays._
+
+- `Error`
+  ```
+  {
+    error: <message>,
+  }
+  ```
+
+- `Direction`
+  ```
+  INCOMING or OUTGOING
+  ```
+
+- `Stream`
+  ```
+  {
+    "localAddr": <multiaddr>,
+    "localID": <peer ID>,
+    "remoteAddr": <multiaddr>,
+    "remoteID": <peer ID>,
+    "createdAt": <time>,
+    "direction": <Direction>,
+    "socketPath": <path>,
```

Comment:
Since we've discussed multi-transport virtualisation, I'd rather make this a map. Some mechanisms like SHM don't deal with sockets, so I'd find an alternate term like "bindings", e.g.

```
"bindings": [
  { "transport": "unix", "addr": "/unixsock/[var/lib/libp2p/links/1234.sock]" },
  { "transport": "shm", "addr": "..." },
]
```

Reply:
like this a lot